pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +212,174 @@ pub mod pallet {
 			todo!()
 		}
 
-		// TODO: not sure if we really need that. Check that out with Raphael.
 		#[pallet::call_index(6)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, Some(who))
 		}
 
 		#[pallet::call_index(7)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn force_start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn force_start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, None)
 		}
 
-		// todo: check if we really need that tx.
 		#[pallet::call_index(8)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn destroy_accounts(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn refund_account(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			account: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let who = T::Lookup::lookup(account)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			let pool_account = pool_id.clone().into();
+
+			let mut total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::ZeroAmount);
+
+			let total_issuances: Vec<(FungiblesAssetIdOf<T>, FungiblesBalanceOf<T>)> = pool_details
+				.bonded_currencies
+				.iter()
+				.map(|id| (id.clone(), T::Fungibles::total_issuance(id.clone())))
+				.filter(|(_, iss)| iss.gt(&Zero::zero()))
+				.collect();
+
+			let mut sum_of_issuances: FungiblesBalanceOf<T> = total_issuances
+				.iter()
+				.fold(Zero::zero(), |sum, (_, x)| sum.saturating_add(*x));
+
+			ensure!(!sum_of_issuances.is_zero(), Error::<T>::ZeroAmount);
+
+			let mut dead_currencies = Vec::with_capacity(total_issuances.len());
+			for (currency_id, token_issuance) in total_issuances.iter() {
+				let burnt = T::Fungibles::decrease_balance(
```

Comment:
`decrease_balance` is not updating the total supply of the coin. Is this intendant? I think what you want is: `burn_from` from the `Mutate` trait.

Reply:
You're right, `burn_from` is what we want - I was just confused by its misleading docstring which suggested it would require you to look up the account's balance first.

Reply:
15369478fd8d5b498cdb7b59f0941bf5bee7fca8